core/mutator/mutator.go

```diff
@@ -52,11 +51,23 @@ type Mutator interface {
 	Mutate(value, mutation []byte) ([]byte, error)
 }
 
+// MutationInfo contains information related mutations, specifically the index
+// and the mutation data.
+type MutationInfo struct {
+	// Index determines the index of the mutation in the map.
+	Index []byte
+	// Data contains the actual mutation data.
+	Data []byte
+}
+
 // Mutation reads and writes mutations to the database.
+// TODO: Add mapID to this interface to support multiple maps per server.
 type Mutation interface {
-	// Read reads all mutations for a specific given mapID, epoch, and index.
-	// TODO(cesarghali): Add mapID to this interface to support multiple maps per server.
-	Read(ctx context.Context, txn transaction.Txn, epoch int64, index []byte) ([]byte, error)
-	// Write saves the mutation in the database.
-	Write(ctx context.Context, txn transaction.Txn, epoch int64, index, mutation []byte) error
+	// ReadRange reads all mutations for a specific given mapID and sequence
+	// range. The range is identified by a starting sequence number and a
+	// count.
+	ReadRange(txn transaction.Txn, startSequence uint64, count int) ([]MutationInfo, error)
+	// Write saves the mutation in the database. Write returns the sequence
+	// number that is written.
+	Write(txn transaction.Txn, index, mutation []byte) (uint64, error)
```

Comment:
It looks like the index is being stored here and inside the mutation itself. 
Consider just passing the mutation itself as the argument to Write. Then the internal implementation can pull out the index if it needs to.  It also saves us a proto serialization too.